File: vega-compose-multiplatform/src/jvmTest/kotlin/dev/aster/vega/compose/mp/CanonicalCalls.kt
package dev.aster.vega.compose.mp

import dev.aster.vega.scene.RasterImage
import kotlin.math.abs
import kotlin.math.floor

/**
 * The sequence of draw calls a scene produces, written so that **two implementations can be
 * compared**.
 *
 * There are two walks over a scene — this module's `SceneWalk` and the Swift package's — and each
 * claims to emit "the same calls in the same order". Nothing checked that. It cost a real defect:
 * the Swift walk had no zero-opacity guard, so on `label-overlap.vg.json` it drew 43 text runs
 * where this one drew 19, and the two were only compared by a person reading two files.
 *
 * `RecordingTarget` cannot do the comparing. There is one on each side and their formats drifted
 * the way anything with two copies does — `->` against ` -> `, `@` against ` a`, a path summarised
 * two different ways — and aligning them would churn every assertion written against either. So
 * this is a *third* recorder, on each side, whose only job is to be byte-identical:
 * `CanonicalCalls` here and `CanonicalCalls.swift` there, one golden file between them.
 *
 * **The format, which is a contract between two languages.** Every field is written, always, in a
 * fixed order, with no shorthand for a default — a recorder that omits what it thinks is
 * uninteresting is a recorder that cannot prove agreement about it. Numbers are three decimals with
 * ties rounded away from zero, spelled out in both languages rather than left to a platform's `%f`:
 * Java's rounds ties up and C's rounds them to even, which is a difference that would show up as a
 * parity failure once in a thousand coordinates and be mystifying.
 *
 * ```
 * group clip=-|(x,y,w,h)
 *   rect (x,y,w,h) corners=[tl,tr,br,bl] fill=<brush> stroke=<stroke>
 *   line (x,y)-(x,y) stroke=<stroke>
 *   path <commands> fill=<brush> stroke=<stroke>
 *   text "…" origin=(x,y) anchor=(x,y) ascent=a font=… size=… weight=… italic=0 angle=0 …
 *   image url="…" raster=-|WxH (x,y,w,h) fit=… smooth=… opacity=…
 * end
 * ```
 *
 * A path is written **command by command**, control points included. The two `RecordingTarget`s
 * both summarise one — a count and a tally here, the first and last there — and a summary is
 * exactly what hides a curve drawn through the wrong control points, which is the kind of thing
 * these two walks could disagree about.
 */
internal class CanonicalCalls : SceneDrawTarget {

  private val lines = mutableListOf<String>()
  private var depth = 0

  val text: String
    get() = lines.joinToString("\n")

  override fun beginGroup(clip: DrawRect?) {
    note("group clip=" + (clip?.let { rect(it) } ?: "-"))
    depth += 1
  }

  override fun endGroup() {
    depth = maxOf(0, depth - 1)
    note("end")
  }

  override fun rect(
    rect: DrawRect,
    corners: DrawCorners,
    fill: DrawBrush?,
    stroke: DrawStroke?,
  ) {
    note(
      "rect ${rect(rect)} corners=[${num(corners.topLeft)},${num(corners.topRight)}," +
        "${num(corners.bottomRight)},${num(corners.bottomLeft)}]" +
        " fill=${brush(fill)} stroke=${stroke(stroke)}"
    )
  }

  override fun line(from: DrawPoint, to: DrawPoint, stroke: DrawStroke?) {
    note("line ${point(from)}-${point(to)} stroke=${stroke(stroke)}")
  }

  override fun path(commands: List<DrawPathCommand>, fill: DrawBrush?, stroke: DrawStroke?) {
    val written =
      commands.joinToString("") {
        when (it) {
          is DrawPathCommand.MoveTo -> "M${point(it.to)}"
          is DrawPathCommand.LineTo -> "L${point(it.to)}"
          is DrawPathCommand.CubicTo -> "C${point(it.control1)}${point(it.control2)}${point(it.to)}"
          DrawPathCommand.Close -> "Z"
        }
      }
    note("path $written fill=${brush(fill)} stroke=${stroke(stroke)}")
  }

  override fun text(run: DrawTextRun, fill: DrawBrush?, stroke: DrawStroke?) {
    note(
      "text ${quoted(run.text)} origin=${point(run.origin)} anchor=${point(run.anchor)}" +
        " ascent=${num(run.ascent)} font=${quoted(run.fontFamily)} size=${num(run.fontSize)}" +
        " weight=${run.fontWeight} italic=${flag(run.italic)}" +
        " angle=${num(run.angleDegrees)} spacing=${num(run.letterSpacing)}" +
        " fill=${brush(fill)} stroke=${stroke(stroke)}"
    )
  }

  override fun image(
    url: String,
    raster: RasterImage?,
    rect: DrawRect,
    fit: DrawImageFit,
    smooth: Boolean,
    opacity: Double,
  ) {
    // A raster's pixels are neither readable nor stable as text, so it is written by its extent.
    // Its
    // *presence* is what the two walks could disagree about.
    note(
      "image url=${quoted(url)} raster=" +
        (raster?.let { "${it.width}x${it.height}" } ?: "-") +
        " ${rect(rect)} fit=${fit.name.lowercase()} smooth=${flag(smooth)}" +
        " opacity=${num(opacity)}"
    )
  }

  private fun note(line: String) {
    lines += "  ".repeat(depth) + line
  }

  private fun flag(value: Boolean) = if (value) "1" else "0"

  private fun point(point: DrawPoint) = "(${num(point.x)},${num(point.y)})"

  private fun rect(rect: DrawRect) =
    "(${num(rect.x)},${num(rect.y)},${num(rect.width)},${num(rect.height)})"

  private fun brush(brush: DrawBrush?): String =
    when (brush) {
      null -> "-"
      is DrawBrush.Solid -> paint(brush.paint)
      is DrawBrush.Linear -> "linear${point(brush.from)}-${point(brush.to)}[${stops(brush.stops)}]"
      is DrawBrush.Radial ->
        "radial${point(brush.center)}r=${num(brush.radius)}[${stops(brush.stops)}]"
    }

  private fun stops(stops: List<DrawStop>) =
    stops.joinToString(",") { "${num(it.offset)}=${paint(it.paint)}" }

  private fun stroke(stroke: DrawStroke?): String =
    if (stroke == null) "-"
    else
      "${brush(stroke.brush)} w=${num(stroke.width)} cap=${stroke.cap.name.lowercase()}" +
        " join=${stroke.join.name.lowercase()} miter=${num(stroke.miterLimit)}" +
        " dash=[${stroke.dash.joinToString(",") { num(it) }}]" +
        " dashOffset=${num(stroke.dashOffset)}"

  private fun paint(paint: DrawPaint): String =
    "#" +
      listOf(paint.red, paint.green, paint.blue, paint.alpha).joinToString("") { channel ->
        val byte = halfAwayFromZero(channel.coerceIn(0.0, 1.0) * 255.0).toInt()
        byte.toString(16).padStart(2, '0')
      }

  /** JSON-ish quoting, so a label containing a quote or a newline cannot break a line. */
  private fun quoted(value: String): String = buildString {
    append('"')
    for (character in value) {
      when (character) {
        '"' -> append("\\\"")
        '\\' -> append("\\\\")
        '\n' -> append("\\n")
        '\r' -> append("\\r")
        '\t' -> append("\\t")
        else -> append(character)
      }
    }
    append('"')
  }

  /**
   * Three decimals, ties away from zero, `-0` written as `0`.
   *
   * Spelled out rather than handed to a format string: Java's `%.3f` rounds a tie up and C's rounds
   * it to even, so `String.format` here and `String(format:)` there would disagree about a
   * coordinate ending in exactly `…5` — a parity failure once in a thousand numbers, with no way to
   * read it.
   */
  private fun num(value: Double): String {
    if (value.isNaN()) return "nan"
    if (value.isInfinite()) return if (value > 0) "inf" else "-inf"
    val thousandths = halfAwayFromZero(value * 1000.0)
    val negative = thousandths < 0
    val magnitude = abs(thousandths)
    val whole = (magnitude / 1000).toLong()
    val fraction = (magnitude % 1000).toLong()
    val sign = if (negative && (whole != 0L || fraction != 0L)) "-" else ""
    return "$sign$whole.${fraction.toString().padStart(3, '0')}"
  }

  private fun halfAwayFromZero(value: Double): Double =
    if (value >= 0.0) floor(value + 0.5) else -floor(-value + 0.5)
}

File: swift/AsterVegaRender/Tests/AsterVegaRenderTests/CanonicalCalls.swift
import AsterVega

@testable import AsterVegaRender

/// The sequence of draw calls a scene produces, written so that **two implementations can be
/// compared**.
///
/// The mirror of `CanonicalCalls.kt` in `vega-compose-multiplatform`, and the format is a contract
/// between the two: same fields, same order, same rounding, byte for byte. Its whole purpose is that
/// the goldens in `test-fixtures/scene-walk/` can be written by one walk and asserted by the other.
///
/// There are two walks over a scene and each is documented as emitting "the same calls in the same
/// order". Nothing checked it, and it cost a defect that shipped: this walk had no zero-opacity
/// guard, so on `label-overlap.vg.json` it emitted 43 text runs where the Compose walk emitted 19.
/// Both renderers' own tests passed throughout, each asserting about itself.
///
/// `RecordingTarget` cannot do the comparing. There is one on each side, their formats drifted the way
/// anything with two copies does, and aligning them would churn every assertion written against
/// either. So this is a third recorder whose only job is to be identical to its counterpart.
///
/// Every field is written, always, with no shorthand for a default: a recorder that omits what it
/// thinks is uninteresting cannot prove agreement about it. Numbers are three decimals with ties
/// rounded **away from zero**, spelled out rather than handed to `String(format:)` — C rounds a tie to
/// even and Java rounds it up, which would show up as a parity failure once in a thousand coordinates
/// and be unreadable.
struct CanonicalCalls: DrawTarget {

  private var lines: [String] = []
  private var depth = 0

  var text: String { lines.joined(separator: "\n") }

  mutating func beginGroup(clip: Rect?) {
    note("group clip=" + (clip.map { show($0) } ?? "-"))
    depth += 1
  }

  mutating func endGroup() {
    depth = max(0, depth - 1)
    note("end")
  }

  mutating func rect(
    _ rect: Rect, corners: AsterVegaRender.Corners, fill: Brush?, stroke: StrokePaint?,
    blend: SceneBlendMode
  ) {
    note(
      "rect \(show(rect)) corners=[\(num(corners.topLeft)),\(num(corners.topRight)),"
        + "\(num(corners.bottomRight)),\(num(corners.bottomLeft))]"
        + " fill=\(show(fill)) stroke=\(show(stroke)) blend=\(show(blend))"
    )
  }

  mutating func line(from: Point, to: Point, stroke: StrokePaint?, blend: SceneBlendMode) {
    note("line \(show(from))-\(show(to)) stroke=\(show(stroke)) blend=\(show(blend))")
  }

  mutating func path(
    _ commands: [AsterVegaRender.PathCommand], fill: Brush?, stroke: StrokePaint?,
    blend: SceneBlendMode
  ) {
    let written = commands.map { command -> String in
      switch command {
      case .move(let to): return "M\(show(to))"
      case .line(let to): return "L\(show(to))"
      case .cubic(let one, let two, let to): return "C\(show(one))\(show(two))\(show(to))"
      case .close: return "Z"
      }
    }
    .joined()
    note("path \(written) fill=\(show(fill)) stroke=\(show(stroke)) blend=\(show(blend))")
  }

  mutating func text(
    _ run: DrawTextRun, fill: Brush?, stroke: StrokePaint?, blend: SceneBlendMode
  ) {
    note(
      "text \(quoted(run.text)) origin=\(show(run.origin)) anchor=\(show(run.anchor))"
        + " ascent=\(num(run.ascent)) font=\(quoted(run.fontFamily)) size=\(num(run.fontSize))"
        + " weight=\(run.fontWeight) italic=\(flag(run.italic))"
        + " angle=\(num(run.angleDegrees)) spacing=\(num(run.letterSpacing))"
        + " fill=\(show(fill)) stroke=\(show(stroke)) blend=\(show(blend))"
    )
  }

  mutating func image(
    url: String,
    raster: DrawRaster?,
    in rect: Rect,
    fit: DrawImageFit,
    smooth: Bool,
    opacity: Double,
    blend: SceneBlendMode
  ) {
    note(
      "image url=\(quoted(url)) raster="
        + (raster.map { "\($0.width)x\($0.height)" } ?? "-")
        + " \(show(rect)) fit=\(fit.rawValue) smooth=\(flag(smooth)) opacity=\(num(opacity))"
        + " blend=\(show(blend))"
    )
  }

  private mutating func note(_ line: String) {
    lines.append(String(repeating: "  ", count: depth) + line)
  }

  private func flag(_ value: Bool) -> String { value ? "1" : "0" }

  /// The blend mode's own name, lower-cased, which is what the Kotlin recorder writes.
  ///
  /// Kotlin's enum entries come across as `SceneBlendMode` objects whose `name` is the Kotlin
  /// spelling — `COLOR_DODGE` — so lower-casing it gives one string on both sides.
  private func show(_ blend: SceneBlendMode) -> String { blend.name.lowercased() }

  private func show(_ point: Point) -> String { "(\(num(point.x)),\(num(point.y)))" }

  private func show(_ rect: Rect) -> String {
    "(\(num(rect.x)),\(num(rect.y)),\(num(rect.width)),\(num(rect.height)))"
  }

  private func show(_ brush: Brush?) -> String {
    switch brush {
    case nil: return "-"
    case .solid(let paint): return show(paint)
    case .linear(let from, let to, let stops, _):
      return "linear\(show(from))-\(show(to))[\(show(stops))]"
    case .radial(let centre, let radius, let stops, _):
      return "radial\(show(centre))r=\(num(radius))[\(show(stops))]"
    }
  }

  private func show(_ stops: [AsterVegaRender.GradientStop]) -> String {
    stops.map { "\(num($0.offset))=\(show($0.paint))" }.joined(separator: ",")
  }

  private func show(_ stroke: StrokePaint?) -> String {
    guard let stroke else { return "-" }
    return "\(show(stroke.brush)) w=\(num(stroke.width)) cap=\(stroke.cap.rawValue)"
      + " join=\(stroke.join.rawValue) miter=\(num(stroke.miterLimit))"
      + " dash=[\(stroke.dash.map(num).joined(separator: ","))]"
      + " dashOffset=\(num(stroke.dashOffset))"
  }

  private func show(_ paint: Paint) -> String {
    "#"
      + [paint.red, paint.green, paint.blue, paint.alpha].map { channel -> String in
        let byte = Int(halfAwayFromZero(min(max(channel, 0), 1) * 255))
        return String(byte, radix: 16, uppercase: false).count == 1
          ? "0" + String(byte, radix: 16) : String(byte, radix: 16)
      }
      .joined()
  }

  /// JSON-ish quoting, so a label containing a quote or a newline cannot break a line.
  private func quoted(_ value: String) -> String {
    var out = "\""
    for character in value {
      switch character {
      case "\"": out += "\\\""
      case "\\": out += "\\\\"
      case "\n": out += "\\n"
      case "\r": out += "\\r"
      case "\t": out += "\\t"
      default: out.append(character)
      }
    }
    return out + "\""
  }

  /// Three decimals, ties away from zero, `-0` written as `0`.
  private func num(_ value: Double) -> String {
    if value.isNaN { return "nan" }
    if value.isInfinite { return value > 0 ? "inf" : "-inf" }
    let thousandths = halfAwayFromZero(value * 1000)
    let magnitude = abs(thousandths)
    let whole = Int64(magnitude / 1000)
    let fraction = Int64(magnitude.truncatingRemainder(dividingBy: 1000))
    let sign = thousandths < 0 && (whole != 0 || fraction != 0) ? "-" : ""
    var digits = String(fraction)
    while digits.count < 3 { digits = "0" + digits }
    return "\(sign)\(whole).\(digits)"
  }

  private func halfAwayFromZero(_ value: Double) -> Double {
    value >= 0 ? (value + 0.5).rounded(.down) : -((-value + 0.5).rounded(.down))
  }
}
